Improve `IsJoinSemilatticeDigraph` and similar methods

MTWhyte opened the pull request:
This PR implements an algorithm described in _Free Lattices_ by Ralph Freese et. al. (1995 edition), which determines whether or not a given partial order digraph is a meet/join semilattice, simultaneously computing the table of meets/joins.

There is an existing algorithm for this, added in #59, which I believe is fairly similar to what I have done here. I modified this existing method in order to simultaneously compute the meet/join tables, and compared the two. The existing method and that in this PR were very similar in terms of computation time, this one maybe being slightly faster, which I have discussed with @james-d-mitchell. Another advantage of the method in this PR is that there's an immediate textbook reference to it, which could be useful in the future.

The most significant modification is in the method for the internal function `DIGRAPHS_IsMeetJoinSemilattice`. Around this, the user functions have been altered so that `DIGRAPHS_IsMeetJoinSemilattice` is run only once, computing and storing both the properties and the tables. The functions `PartialOrderDigraphJoinOfVertices` and `PartialOrderDigraphMeetOfVertices` will use the join/meet tables of a join/meet semilattice digraph, if they are known.

<details>
  <summary>Todo list</summary>
  
  ## Todo
- [x] If a digraph has its adjacency matrix, use this to determine whether elements are less than or equal to each other. This can be achieved by using `IsDigraphEdge(D, u, v)` in place of `u in nbs[v]`.

- [x] Add speed timing graphs to PR text or the wiki

- [x] Add tests for new behaviour
- [x] Documentation for `JoinSemilatticeDigraphJoinTable` and `MeetSemilatticeDigraphJoinTable`
- [x] Add reference to _Free Lattices_ Listing 11.9 as a comment above `DIGRAPHS_IsMeetJoinSemilattice`  
</details>

MTWhyte commented on gap/prop.gi:
```diff
@@ -38,58 +38,92 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+  for x in P do
+    tab[x, x] := x;
+    for y in S do
+      T := [];  # EmptyPlist didn't improve speed
```
I meant to remove the comments here from `prop.gi` and add them as GitHub comments. I will remove them later.

MTWhyte commented:
**Error with mutable digraphs (note to self, mainly)**

I have noticed some strange behaviour with mutable digraphs, in what I have written. I get the following:

```
gap> D := Digraph(IsMutable, [[2, 3], [4], [4], []]);
<mutable digraph with 4 vertices, 4 edges>
gap> DigraphReflexiveTransitiveClosure(D);
<mutable digraph with 4 vertices, 9 edges>
gap> JoinSemilatticeDigraphJoinTable(D);
fail
gap> IsJoinSemilatticeDigraph(D);
false
gap> MeetSemilatticeDigraphMeetTable(D);
[ [ 1, 1, 1, 1 ], [ 1, 2, 1, 2 ], [ 1, 1, 3, 3 ], [ 1, 2, 3, 4 ] ]
gap> IsMeetSemilatticeDigraph(D);
true
gap> D := DigraphImmutableCopy(D);
<immutable digraph with 4 vertices, 9 edges>
gap> IsJoinSemilatticeDigraph(D);
true
gap> JoinSemilatticeDigraphJoinTable(D);
[ [ 1, 2, 3, 4 ], [ 2, 2, 4, 4 ], [ 3, 4, 3, 4 ], [ 4, 4, 4, 4 ] ]
```

This indicates something has gone wrong with how mutable digraphs are handles, which I shall investigate.

MTWhyte commented:
### Some work regarding computation times

Here are some graphs regarding the speeds of the methods I tested. The first graph refers to the _New method_, the _Existing adaptation_, and the _Naïve method_. These are described as follows:

- **_New method:_** The algorithm in this PR
- **_Existing adaptation:_** Adapting the old `DIGRAPHS_IsMeetJoinSemilattice`, storing meets/joins when they are found (Code is in the appendix)
- **_Naïve method:_** Constructing meet/join tables of lattices, by iterating through pairs `u` and `v` of vertices, using `PartialOrderDigraphJoinOfVertices(D, u, v)` (and analogously for meets). The commutativity of meets and joins was used, so each pair of vertices as a set was only considered once.

<img width="731" alt="image" src="https://user-images.githubusercontent.com/42549861/160831517-4d1d8f91-34ef-4709-a966-5186e313cba0.png">

The methods seem to have similar complexity. It's worth noting that naïve method does not determine whether or not its argument is a meet/join semilattice, whereas the other two methods do this concurrently with computing the tables. The naïve method in green is really just there for illustration, but the principal question was which of the other two methods is better.

<img width="731" alt="image" src="https://user-images.githubusercontent.com/42549861/160832000-af5b8918-d906-48b8-85c2-281fc1a38276.png">

The above graph compares the time between the new method and the existing adaptation for determining the property _IsJoinSemilatticeDigraph_, but not the tables themselves. The existing adaptation seems marginally faster in this situation, I think because it can avoid adding a lot of stuff to lists in this case, whereas the new method must add a lot of stuff to lists as it computes the join table in the very process of determining whether its argument is a join semilattice.

In discussion with @james-d-mitchell, it was decided that it's better to just automatically compute the meet/join tables for meet/join semilattices in conjunction with determining whether or not they are in fact meet/join semilattices. This makes the first of the above graphs more relevant than the second.

<img width="731" alt="image" src="https://user-images.githubusercontent.com/42549861/160832774-85e662c7-163f-4f30-b8ab-b8ac18f9b015.png">

This graph above runs the same process as the first graph on random lattices, but on a larger range of vertices. It appears that for larger `n`, the new method is marginally faster than the existing adaptation. Running a similar process on random digraphs at large, rather than those which are necessarily lattices, gives the following:

<img width="731" alt="image" src="https://user-images.githubusercontent.com/42549861/160833496-90877b69-5bf2-43d4-8519-568750984839.png">

In all, the new method and the existing adaptation are reasonably similar, with the new method perhaps taking the lead as the number of vertices increases. The claimed time complexity of the new method is `O(n^2.5)`, where `n` is the number of vertices of the digraph. Doing computations for random lattices and random digraphs, I get the following

<img width="731" alt="image" src="https://user-images.githubusercontent.com/42549861/160834367-c2649e1c-9356-4661-b9bb-a5a36dbdc8de.png">
<img width="709" alt="image" src="https://user-images.githubusercontent.com/42549861/160834471-cea5ec00-2bc4-4c0a-94e6-3bc3b7b56f58.png">

In both, I have fitted a line to some of the data where it looks reasonably linear. It being a log-log plot, the gradient of this line should be the complexity. This suggests that the measured time complexity is about 2. It's worth noting that this was just a rough measurement: `RandomDigraph(n)` returns a digraph with `n` vertices, but `RandomLattice(n)` returns a digraph with somewhere between `n` and `2 n` vertices. I've plotted this as if `n` is the number of vertices, but since the two are reasonably proportional it shouldn't be that important.



#### Appendix: Code for the _Existing adaptation_

```
DIGRAPHS_IsMeetJoinSemilattice2 := function(nbs, topo)
  local i, j, k, n, x, len, table;

  # The IsHomogenousList assumes the list <nbs> contains no or only empty lists
  n := Length(nbs);
  if topo <> false then
    table := List([1 .. n], x -> []);
  fi;
  for i in [1 .. n] do
    if topo <> false then
      table[topo[i]][topo[i]] := topo[i];
    fi;
    for j in [i + 1 .. n] do
      if j in nbs[i] or i in nbs[j] then
        if topo <> false then
          if j in nbs[i] then
            table[topo[i]][topo[j]] := topo[j];
            table[topo[j]][topo[i]] := topo[j];
          else
            table[topo[i]][topo[j]] := topo[i];
            table[topo[j]][topo[i]] := topo[i];
          fi;
        fi;
        continue;
      fi;
      x := Intersection(nbs[i], nbs[j]);
      if IsEmpty(x) then
        return false;
      fi;
      len := Length(x);
      k   := x[len];  # Check whether <k> is the meet of <i> and <j>
      if Length(nbs[k]) < len then
        if topo <> false then
          return fail;
        fi;
        return false;
      fi;
      if topo <> false then
        table[topo[i]][topo[j]] := topo[k];
        table[topo[j]][topo[i]] := topo[k];
      fi;
    od;
  od;
  if topo <> false then
    return table;
  fi;
  return true;
end;

IsJoinSemilattice2 := function(D)
  local topo;
  if not IsPartialOrderDigraph(D) then
    return false;
  fi;
  topo := DigraphTopologicalSort(D);
  D    := DigraphMutableCopy(D);
  D    := OnDigraphs(D, PermList(topo) ^ -1);
  Apply(D!.OutNeighbours, Set);
  return DIGRAPHS_IsMeetJoinSemilattice2(D!.OutNeighbours, false);
end;

JoinTable2 := function(D)
  local topo;
  if not IsPartialOrderDigraph(D) then
    return false;
  fi;
  topo := DigraphTopologicalSort(D);
  D    := DigraphMutableCopy(D);
  D    := OnDigraphs(D, PermList(topo) ^ -1);
  Apply(D!.OutNeighbours, Set);
  return DIGRAPHS_IsMeetJoinSemilattice2(D!.OutNeighbours, topo);
end;
```

MTWhyte commented on gap/prop.gi:
```diff
@@ -38,58 +38,99 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+  for x in P do
+    tab[x, x] := x;
+    for y in S do
+      T := [];
```
Using `EmptyPlist` didn't improve speed

MTWhyte commented on gap/prop.gi:
```diff
@@ -38,58 +38,99 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+    for y in S do
+      T := [];
+      for z in U[x] do
+        Add(T, tab[y, z]);
```
Using `SetAdd` didn't improve speed

james-d-mitchell commented on gap/oper.gi:
```diff
@@ -2187,7 +2192,12 @@ InstallMethod(PartialOrderDigraphMeetOfVertices,
+
+  if HasMeetSemilatticeDigraphMeetTable(D) then
+    tab := MeetSemilatticeDigraphMeetTable(D);
+    return tab[i, j];
```
```suggestion
    return MeetSemilatticeDigraphMeetTable(D)[i, j];
```

james-d-mitchell commented on gap/attr.gi:
```diff
@@ -2675,3 +2675,27 @@ function(D)
   return DIGRAPHS_NonSemimodularPair(InNeighbours(D)) = fail;
 end);
+
+InstallMethod(JoinSemilatticeDigraphJoinTable, "for a digraph",
```
I completely understand the rationale for the name, but I think `DigraphJoinTable` would better!

MTWhyte replied:
Done

james-d-mitchell commented on gap/prop.gi:
```diff
@@ -38,58 +38,99 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+    return false;
+  fi;
+
+  D   := DigraphImmutableCopyIfMutable(D);
```
```suggestion
  D   := DigraphMutable(D);
```

MTWhyte replied:
I think this is replaced by what we discussed later on

james-d-mitchell commented on gap/prop.gi:
```diff
@@ -38,58 +38,99 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+  D   := DigraphImmutableCopyIfMutable(D);
+
+  if IsMultiDigraph(D) then
+    D := DigraphRemoveAllMultipleEdges(D);
```
```suggestion
    DigraphRemoveAllMultipleEdges(D);
```

MTWhyte commented on gap/prop.gi:
```diff
@@ -38,58 +38,99 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
-      if Length(nbs[k]) < len then
+# The method below is based on Listing 11.9 of 'Free Lattices'
+# by Ralph Freese et. al.
+InstallMethod(DIGRAPHS_IsMeetJoinSemilatticeDigraph, "for a digraph and a bool",
```
Done

MTWhyte commented on gap/prop.gi:
```diff
@@ -38,60 +38,103 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+  P      := Reversed(DigraphTopologicalSort(D));
+  U      := InNeighbours(DigraphReflexiveTransitiveReduction(copy));
+  tab    := DIGRAPHS_MeetJoinTable(D, P, U, false);
+  SetDigraphMeetTable(D, tab);
```
Check that `D` is immutable first.

MTWhyte replied:
Done

MTWhyte commented on gap/prop.gi:
```diff
@@ -38,60 +38,103 @@ D -> IsConnectedDigraph(D) and IsEmpty(ArticulationPoints(D)));
+  P      := DigraphTopologicalSort(D);
+  U      := OutNeighbours(DigraphReflexiveTransitiveReduction(copy));
+  tab    := DIGRAPHS_MeetJoinTable(D, P, U, true);
+  SetDigraphJoinTable(D, tab);
```
Done

MTWhyte commented:
*Testing note*

For good measure in testing the correctness of this work, I have compared it to more elementary methods for computing the meet and join tables of lattices, and found them to give the same results on all 37622 lattices on 11 vertices. The elementary functions described are as follows:

```
JoinTable := function(D)
  local N, tab, i, j;
  N   := DigraphNrVertices(D);
  tab := List([1 .. N], x -> []);
  for i in [1 .. N] do
    for j in [i .. N] do
      tab[i][j] := PartialOrderDigraphJoinOfVertices(D, i, j);
      tab[j][i] := tab[i][j];
    od;
  od;
  return tab;
end;

MeetTable := function(D)
  local N, tab, i, j;
  N   := DigraphNrVertices(D);
  tab := List([1 .. N], x -> []);
  for i in [1 .. N] do
    for j in [i .. N] do
      tab[i][j] := PartialOrderDigraphMeetOfVertices(D, i, j);
      tab[j][i] := tab[i][j];
    od;
  od;
  return tab;
end;
```

This was run on the current `master`, so before any changes to `PartialOrderDigraphJoinOfVertices` and `PartialOrderDigraphMeetOfVertices`.

MTWhyte commented on doc/z-chap5.xml:
```diff
@@ -28,6 +28,7 @@
     <#Include Label="IsPreorderDigraph">
     <#Include Label="IsPartialOrderDigraph">
     <#Include Label="IsMeetSemilatticeDigraph">
+    <#Include Label="DigraphMeetTable">
```
I'm not actually sure. I understand that `DigraphMeetTable` is the 'label' for the documentation entry discussing both `DigraphMeetTable` and `DigraphJoinTable`. I've done this based on how `IsLatticeDigraph`, `IsMeetSemilatticeDigraph` and `IsJoinSemilattice` digraph are documented. Calling these up in the documentation gives
```
6.3-3 IsMeetSemilatticeDigraph

  ‣ IsMeetSemilatticeDigraph( digraph ) ───────────────────────────── property
  ‣ IsJoinSemilatticeDigraph( digraph ) ───────────────────────────── property
  ‣ IsLatticeDigraph( digraph ) ───────────────────────────────────── property
  Returns:  true or false.
```
In the same sort of style, just having `DigraphMeetTable` point at the documentation for both the meet and join table gives
```
6.3-4 DigraphMeetTable

  ‣ DigraphMeetTable( digraph ) ──────────────────────────────────── attribute
  ‣ DigraphJoinTable( digraph ) ──────────────────────────────────── attribute
  Returns:  A matrix or fail.
```

I don't know if this is optimal, but I think it's how to get the same style.

flsmith replied:
Ah, I see.

flsmith commented on tst/standard/attr.tst:
```diff
@@ -2804,6 +2804,36 @@ false
+[ [ 1, 2, 3, 3 ], [ 2, 2, 3, 3 ], [ 3, 3, 3, 3 ], [ 3, 3, 3, 4 ] ]
+gap> DigraphMeetTable(D);
+fail
+
```
One last comment @MTWhyte - it would be nice to have these tests also explicitly check the known value of these joins/meets.

MTWhyte commented:
Thanks @flsmith, I've added explicit meet/join checking now. It occurred to me that, for the larger example I added, the explicit checking I was doing was in the wrong order. Since this PR makes `PartialOrderDigraphJoinOfVertices` look up the appropriate entry of `DigraphJoinTable` if it is known, I've changed it to compute the test meets/joins before the table, so the other method is used. (I guess this doesn't matter for the mutable example, but I think it makes more sense to keep everything in the same 'style'.)

james-d-mitchell commented:
@wilfwilson any ideas why the cygwin job is failing? Seems to be an issue with the CI rather than with this PR.

wilfwilson commented:
Yes it's a problem with `gap-actions/setup-gap` as linked in the GitHub interface above this comment. I've tried something and hoping it works.

wilfwilson commented:
@james-d-mitchell It's getting further than it was before, hopefully it's sorted.
